### PR TITLE
Improve logging of the sidecar

### PR DESCRIPTION
**Pull request by bwoebi**
This allows dynamically updating the log level of the sidecar as well as tracking logs from telemetry and the sidecar operation itself in trace mode.

Further, given --cfg tokio-taskdump, the sidecar is able to dump a task snapshot of itself.

Notes: telemetry required the tracing-subscriber, but it's only for the example. Moved it to dev-dependencies.

### CODE REVIEW

File: sidecar/src/setup/unix.rs [hunk 1]
```diff
@@ -58,15 +63,15 @@ impl Liaison for SharedDirLiaison {
             // failing to acquire lock
             // means that another process is creating the socket
             Err(err) => {
-                tracing::debug!("failed_locking");
+                debug!("failed_locking");
```
Comment (morrisonlevi) [inline sidecar/src/setup/unix.rs:66]: Maybe shift this to `warn!`? Seems more important than some of the other debugs like "already_listening" which should be fine.

File: sidecar/src/interface.rs [hunk 2]
```diff
@@ -682,7 +726,9 @@ impl SidecarServer {
             None => return,
         };
 
-        session.shutdown().await
+        info!("Shutting down session: {}", session_id);
+        session.shutdown().await;
+        debug!("Successfully shut down session: {}", session_id);
```
Comment (morrisonlevi) [inline sidecar/src/interface.rs:731]: It seems like maybe the first item should be `debug!` and the second should be `info!`?

**Review by morrisonlevi — COMMENTED** [review]
Overall it looks good. I have some suggestions for possible changes to log levels, and this is minor of course.

I have a more serious question, though: why are only some things guarded behind `cfg(tracing)`? Seems like it might break if you build without the feature, but maybe they are protected somehow by code that's not in the diff.

**Comment by bwoebi** [comment]
It's only the tracing_subscriber which must be guarded, everything else is always present.

File: sidecar/src/interface.rs (hunk 2 above)
Comment (bwoebi) [reply sidecar/src/interface.rs:731]: The intention was noting the moment in time where no new data is accepted for the session. The second is really just "successfully cleaned up".

File: sidecar/src/log.rs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    let log_file = std::fs::File::options()
+        .create(true)
+        .truncate(false)
+        .write(true)
+        .append(true)
+        .open(path)?;
+    Ok(log_file)
+}
+
+pub struct TemporarilyRetainedMap<K, V>
```
Comment (bantonsson) [inline sidecar/src/log.rs:38]: Maybe the code and the logic in this file is self explanatory to everyone else, but I would welcome some comments outlining how things fit together. Especially the interaction between this code and the code in `tracing` that calls back into this code to compute the log levels.

Also, are there any tests for this functionality?

Comment (bwoebi) [reply sidecar/src/log.rs:38]: I agree, let me add a couple comments.

The sidecar code is mostly tested through tests from the dd-trace-php repository. I haven't written self-contained tests for the sidecar yet.

Comment (bantonsson) [reply sidecar/src/log.rs:40]: I would really like something as complex as a reference counting data structure (with the potential to leak memory) to have some unit tests.

**Comment by bwoebi** [comment]
@bantonsson Added a test :-) Looks good now?

**Review by bantonsson — APPROVED** [review]

**Pull request merged**